src/btree/bt_npos.c

```diff
@@ -0,0 +1,352 @@
+    uint32_t entries, slot;
+    int unused = 1; /* WT_UNUSED(snprintf) is cooked in GCC */
+
+    npos = start;
+    if (path_str)
+        *path_str_offsetp = 0;
+
+    WT_ENTER_PAGE_INDEX(session);
+    while (!__wt_ref_is_root(ref)) {
+        slot = UINT32_MAX; /* We get this invalid value in case of error */
+        __wt_ref_index_slot(session, ref, &pindex, &slot);
+        entries = pindex->entries;
+        /*
+         * Depending on the implementation, `__wt_ref_index_slot` might return an error or `slot`
+         * outside of range. Check for `slot < entries` ensures that it's a valid number. If it's
+         * not, then just skip the adjustment: the resulting number will be wrong but still within
+         * the range of the current page. Alternatively, could assign it any "reasonable" estimate
+         * like 0.5 or 0 / 1 depending on the walk direction.
+         */
+        if (slot < entries)
+            npos = (slot + npos) / entries;
+        if (path_str)
+            WT_UNUSED(unused = __wt_snprintf_len_incr(&path_str[*path_str_offsetp],
+                        path_str_sz_max - *path_str_offsetp, path_str_offsetp,
+                        "[%" PRIu32 "/%" PRIu32 "]", slot, entries));
+        __wt_ref_ascend(session, &ref, NULL, NULL);
+    }
+    WT_LEAVE_PAGE_INDEX(session);
+
+    if (path_str) {
+        path_str[*path_str_offsetp] = 0;
+    }
+
+    return (WT_CLAMP(npos, 0.0, 1.0));
+}
+
+/*
+ * !!!
+ * __find_closest_leaf --
+ *     Find the closest suitable page according to flags.
+ *     - It should not be deleted.
+ *     - If WT_READ_CACHE is set, the page should be in memory.
+ *     - If the initial ref is to a good page, it will be returned.
+ *     - If the initial ref is null, it does nothing.
+ */
+static int
+__find_closest_leaf(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags)
+{
+    if (*refp == NULL || F_ISSET(*refp, WT_REF_FLAG_LEAF))
+        return (0);
+    LF_SET(WT_READ_SKIP_INTL);
+    return (__wt_tree_walk(session, refp, flags));
+}
+
+/*
+ * __page_from_npos_internal --
+ *     Go to a leaf page given its normalized position. Note that this function can return a "bad"
+ *     page (deleted, locked, etc). The caller of this function should walk the tree to find a
+ *     suitable page.
+ *
+ * NOTE: Must be called within WT_WITH_PAGE_INDEX or WT_ENTER_PAGE_INDEX
+ */
+static int
+__page_from_npos_internal(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags, double npos)
+{
+    WT_BTREE *btree;
+    WT_DECL_RET;
+    WT_PAGE *page;
+    WT_PAGE_INDEX *pindex;
+    WT_REF *current, *descent;
+    double npos_local;
+    int idx, entries;
+    bool eviction;
+
+    *refp = NULL;
+
+    btree = S2BT(session);
+    current = NULL;
+    /*
+     * This function is called by eviction to find a page in the cache. That case is indicated by
+     * the WT_READ_CACHE flag. Ordinary lookups in a tree will read pages into cache as needed.
+     */
+    eviction = LF_ISSET(WT_READ_CACHE);
```

Comment:
What is the use case for not eviction? I think we can rename this just to read cache? It is not necessary couple the implementation to eviction.